onnxruntime/core/session/ep_plugin_provider_interfaces.cc

```diff
@@ -0,0 +1,248 @@
+                            << "have an unsupported node in any path between two of the supported nodes.";
+        return {};
+      }
+
+      for (auto& capability : capabilities) {
+        result.push_back(std::move(capability));
+      }
+    } else {
+      LOGS_DEFAULT(ERROR) << "PluginExecutionProvider::GetCapability() has invalid NodeGroupingKind: "
+                          << static_cast<int>(node_grouping.kind);
+      return {};
+    }
+  }
+
+  return result;
+}
+
+PluginExecutionProvider::FusedNodeState& PluginExecutionProvider::PushFusedNodeState(size_t num_fused_nodes) {
+  fused_node_states_.push_back(FusedNodeState());
+  FusedNodeState& fused_node_state = fused_node_states_.back();
+  fused_node_state.nodes.reserve(num_fused_nodes);
+  return fused_node_state;
+}
+
+EpNode& PluginExecutionProvider::FusedNodeState::AddFusedNode(const Node& fused_node) {
+  auto ep_fused_node = EpNode::Create(fused_node, /*parent graph*/ nullptr, this->value_infos);
+  this->nodes.push_back(std::move(ep_fused_node));
+  return *this->nodes.back();
+}
+
+common::Status PluginExecutionProvider::Compile(const std::vector<FusedNodeAndGraph>& fused_nodes_and_graphs,
+                                                std::vector<NodeComputeInfo>& node_compute_infos) {
+  const logging::Logger* logger = GetLogger();
+  const size_t num_graphs = fused_nodes_and_graphs.size();
+  std::vector<std::unique_ptr<EpGraph>> api_graphs_holder;
+  std::vector<const OrtGraph*> api_graphs;
+  std::vector<OrtNodeComputeInfo*> api_node_compute_infos(num_graphs, nullptr);
+  std::vector<const OrtNode*> api_fused_nodes;
+  FusedNodeState& fused_node_state = PushFusedNodeState(num_graphs);
+
+  api_graphs_holder.reserve(num_graphs);
+  api_graphs.reserve(num_graphs);
+  api_fused_nodes.reserve(num_graphs);
+
+  // Wrap GraphViewers into OrtGraphs and fused Nodes into OrtNodes.
+  for (const FusedNodeAndGraph& node_and_graph : fused_nodes_and_graphs) {
+    const GraphViewer& graph_viewer = node_and_graph.filtered_graph;
+    const Node& fused_node = node_and_graph.fused_node;
+
+    auto ep_graph = EpGraph::Create(graph_viewer);
+    api_graphs.push_back(ep_graph->ToExternal());
+    api_graphs_holder.push_back(std::move(ep_graph));
+
+    EpNode& ep_fused_node = fused_node_state.AddFusedNode(fused_node);
+    api_fused_nodes.push_back(ep_fused_node.ToExternal());
+  }
+
+  ORT_RETURN_IF_ERROR(ToStatus(ort_ep_->Compile(ort_ep_.get(), api_graphs.data(), api_fused_nodes.data(), num_graphs,
+                                                api_node_compute_infos.data())));
+
+  // Save OrtNodeComputeInfo created by OrtEp instance. They're freed when this IExecutionProvider
+  // is destroyed.
+  api_node_compute_infos_.reserve(api_node_compute_infos_.size() + num_graphs);
+  for (size_t i = 0; i < num_graphs; i++) {
+    if (api_node_compute_infos[i] != nullptr) {
+      api_node_compute_infos_.push_back(api_node_compute_infos[i]);
```

Comment:
we should try to avoid the possibility of exceptions between the call to `Compile()` and where the outputs are stored in the managing container. it may be better to move the potentially throwing `vector::reserve` before `Compile`.

Reply:
Moved the reserve call before the call to `Compile`.